File: packages/react-native/ios/Utils/RJKeychainManager.m
//
//  RJKeychainManager.m
//  Rejourney
//
//  Secure Keychain storage implementation.
//
//  Copyright (c) 2026 Rejourney
//

#import "RJKeychainManager.h"
#import "../Core/RJLogger.h"
#import <Security/Security.h>

static NSString *const kRJKeychainServiceName = @"com.rejourney.sdk";

@implementation RJKeychainManager

#pragma mark - Singleton

+ (instancetype)sharedManager {
  static RJKeychainManager *instance = nil;
  static dispatch_once_t onceToken;
  dispatch_once(&onceToken, ^{
    instance = [[self alloc] init];
  });
  return instance;
}

#pragma mark - Query Building

- (NSMutableDictionary *)baseQueryForKey:(NSString *)key {
  return [@{
    (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
    (__bridge id)kSecAttrService : kRJKeychainServiceName,
    (__bridge id)kSecAttrAccount : key,
    (__bridge id)kSecAttrAccessible :
        (__bridge id)kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
  } mutableCopy];
}

#pragma mark - Storage Operations

- (BOOL)setString:(NSString *)value forKey:(NSString *)key {
  if (!value || !key)
    return NO;

  
  [self deleteValueForKey:key];

  NSData *valueData = [value dataUsingEncoding:NSUTF8StringEncoding];
  if (!valueData)
    return NO;

  NSMutableDictionary *query = [self baseQueryForKey:key];
  query[(__bridge id)kSecValueData] = valueData;

  OSStatus status = SecItemAdd((__bridge CFDictionaryRef)query, NULL);

  if (status == errSecSuccess) {
    RJLogDebug(@"Keychain: Stored value for key '%@'", key);
    return YES;
  } else {
    RJLogWarning(@"Keychain: Failed to store value for key '%@' (status: %d)",
                 key, (int)status);
    return NO;
  }
}

- (nullable NSString *)stringForKey:(NSString *)key {
  if (!key)
    return nil;

  NSMutableDictionary *query = [self baseQueryForKey:key];
  query[(__bridge id)kSecReturnData] = @YES;
  query[(__bridge id)kSecMatchLimit] = (__bridge id)kSecMatchLimitOne;

  CFDataRef dataRef = NULL;
  OSStatus status = SecItemCopyMatching((__bridge CFDictionaryRef)query,
                                        (CFTypeRef *)&dataRef);

  if (status == errSecSuccess && dataRef) {
    NSData *data = (__bridge_transfer NSData *)dataRef;
    return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
  }

  return nil;
}

- (BOOL)deleteValueForKey:(NSString *)key {
  if (!key)
    return NO;

  NSMutableDictionary *query = [self baseQueryForKey:key];
  OSStatus status = SecItemDelete((__bridge CFDictionaryRef)query);

  return status == errSecSuccess || status == errSecItemNotFound;
}

- (void)clearAll {
  NSDictionary *query = @{
    (__bridge id)kSecClass : (__bridge id)kSecClassGenericPassword,
    (__bridge id)kSecAttrService : kRJKeychainServiceName
  };

  OSStatus status = SecItemDelete((__bridge CFDictionaryRef)query);
  if (status == errSecSuccess) {
    RJLogDebug(@"Keychain: Cleared all Rejourney items");
  }
}

@end

File: packages/react-native/ios/Utils/RJKeychainManager.swift
import Foundation
import Security

@objcMembers
public class RJKeychainManager: NSObject {
    
    public static let shared = RJKeychainManager()
    
    private let serviceName = "com.rejourney.sdk"
    
    private override init() {
        super.init()
    }
    
    // MARK: - Query Builder
    
    private func baseQuery(key: String) -> [String: Any] {
        return [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: serviceName,
            kSecAttrAccount as String: key,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
        ]
    }
    
    // MARK: - Public API
    
    @discardableResult
    public func setString(_ value: String, forKey key: String) -> Bool {
        guard let data = value.data(using: .utf8) else { return false }
        
        let query = baseQuery(key: key)
        
        // Delete existing item first to ensure upsert behavior
        SecItemDelete(query as CFDictionary)
        
        var newQuery = query
        newQuery[kSecValueData as String] = data
        
        let status = SecItemAdd(newQuery as CFDictionary, nil)
        
        if status == errSecSuccess {
            RJLogger.logDebugMessage("Keychain: Stored value for key '\(key)'")
            return true
        } else {
            RJLogger.logErrorMessage("Keychain: Failed to store value for key '\(key)' (status: \(status))")
            return false
        }
    }
    
    public func stringForKey(_ key: String) -> String? {
        var query = baseQuery(key: key)
        query[kSecReturnData as String] = true
        query[kSecMatchLimit as String] = kSecMatchLimitOne
        
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        
        if status == errSecSuccess, let data = result as? Data, let value = String(data: data, encoding: .utf8) {
            return value
        }
        
        return nil
    }
    
    @discardableResult
    public func deleteValue(forKey key: String) -> Bool {
        let query = baseQuery(key: key)
        let status = SecItemDelete(query as CFDictionary)
        return status == errSecSuccess || status == errSecItemNotFound
    }
    
    public func clearAll() {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: serviceName
        ]
        
        let status = SecItemDelete(query as CFDictionary)
        if status == errSecSuccess {
            RJLogger.logDebugMessage("Keychain: Cleared all Rejourney items")
        }
    }
}
